Sites Logout Automatically After Brief Amount of Time

the-chenergy opened the pull request:
#### a2efc000398df38fa2fa081a89da4dc038f4d8cc
<pre>
Sites Logout Automatically After Brief Amount of Time
<a href="https://rdar.apple.com/99829958">rdar://99829958</a>
<a href="https://bugs.webkit.org/show_bug.cgi?id=245491">https://bugs.webkit.org/show_bug.cgi?id=245491</a>

Reviewed by Sihui Liu.

Web Inspector currently uses its own default website data store, one
that comes with WKWebView with the default configurations. However,
this can cause conflicts when the owner browser already created a data
store pointing to the same file paths. As the browser may not expect
the default data store to be used at all, this patch makes Web Inspector
try to use the same custom data store object as the browser, avoiding
the conflicts.

I was not able reproduce the issue from the bug report, but I tested
that this patch still preserved Web Inspector&apos;s storage as normal.
I could still add breakpoints and change settings and have Web Inspector
remember those states across relaunches of Safari or the Minibrowser,
all while my code changes were executed.

* Source/WebKit/UIProcess/Inspector/mac/WKInspectorViewController.mm:
(-[WKInspectorViewController webViewConfiguration]):
   - When constructing the configurations to create a WKWebView, try
     searching for the data store object the browser would have created.

Canonical link: <a href="https://commits.webkit.org/286436@main">https://commits.webkit.org/286436@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/20232a93620f8e14156bb52fcc403833732423fd

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/75727 "Passed style check") | [⏳ 🛠 ios ](https://ews-build.webkit.org/#/builders/iOS-17-Build-EWS "Waiting in queue, processing has not started yet") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/28578 "Built successfully") | [  ~~🛠 wpe~~](https://ews-build.webkit.org/#/builders/5/builds/80216 "Build was cancelled. Recent messages:Printed configuration; Skipping applying patch since patch_id isn't provided; Checked out pull request; Compiled WebKit (warnings)") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/26991 "Built successfully") 
| | [⏳ 🛠 ios-sim ](https://ews-build.webkit.org/#/builders/iOS-17-Simulator-Build-EWS "Waiting in queue, processing has not started yet") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/3015 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/5/builds/80216 "Build was cancelled. Recent messages:Printed configuration; Skipping applying patch since patch_id isn't provided; Checked out pull request; Compiled WebKit (warnings)") | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/17571 "Build is in progress. Recent messages:Printed configuration") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/78794 "Passed tests") | [⏳ 🧪 ios-wk2 ](https://ews-build.webkit.org/#/builders/iOS-17-Simulator-Build-EWS "Waiting in queue, processing has not started yet") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/18/builds/65057 "Build was cancelled. Recent messages:OS: Ventura (13.6.7), Xcode: 14.3; Running apply-patch; Checked out pull request; run-api-tests (cancelled)") | [  ~~🧪 api-wpe~~](https://ews-build.webkit.org/#/builders/5/builds/80216 "Build was cancelled. Recent messages:Printed configuration; Skipping applying patch since patch_id isn't provided; Checked out pull request; Compiled WebKit (warnings)") | 
| | [⏳ 🧪 ios-wk2-wpt ](https://ews-build.webkit.org/#/builders/iOS-17-Simulator-Build-EWS "Waiting in queue, processing has not started yet") | | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/25320 "Build is in progress. Recent messages:Printed configuration") | 
| | [⏳ 🧪 api-ios ](https://ews-build.webkit.org/#/builders/iOS-17-Simulator-Build-EWS "Waiting in queue, processing has not started yet") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/22883 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/81686 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/3066 "Built successfully") | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/1949 "Build is in progress. Recent messages:OS: Sequoia (15.0.1), Xcode: 16.0") | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/67632 "Build is in progress. Recent messages:Printed configuration") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/3217 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/65024 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/66934 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/16721 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/10883 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/9020 "Passed tests") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/3023 "Built successfully") | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/5861 "Build is in progress. Recent messages:OS: Sequoia (15.0.1), Xcode: 16.0") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/3048 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/3983 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/3055 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

szewai commented on Source/WebKit/UIProcess/Inspector/mac/WKInspectorViewController.mm:
```diff
@@ -176,7 +178,18 @@ - (WKWebViewConfiguration *)webViewConfiguration
-    
+
+    // Prefer using a custom persistent data store if one exists.
+    WKWebsiteDataStore *targetDataStore = nil;
```
This should be `RetainPtr`

cdumez replied:
No `= nil` when you switch to `RetainPtr<WKWebsiteDataStore>`

szewai commented on Source/WebKit/UIProcess/Inspector/mac/WKInspectorViewController.mm:
```diff
@@ -176,7 +178,18 @@ - (WKWebViewConfiguration *)webViewConfiguration
+        if (dataStore.isPersistent()
+            && dataStore.sessionID() != PAL::SessionID::defaultSessionID()
+            && !dataStore.configuration().identifier())
+            targetDataStore = WebKit::wrapper(dataStore);
```
Let's compare path instead, and debug assert that `targetDataStore` is only set one

cdumez commented on Source/WebKit/UIProcess/Inspector/mac/WKInspectorViewController.mm:
```diff
@@ -176,7 +178,18 @@ - (WKWebViewConfiguration *)webViewConfiguration
+
+    // Prefer using a custom persistent data store if one exists.
+    WKWebsiteDataStore *targetDataStore = nil;
+    WebKit::WebsiteDataStore::forEachWebsiteDataStore([&targetDataStore](WebKit::WebsiteDataStore &dataStore) {
```
`&` on the wrong side.

cdumez commented on Source/WebKit/UIProcess/Inspector/mac/WKInspectorViewController.mm:
```diff
@@ -176,7 +178,18 @@ - (WKWebViewConfiguration *)webViewConfiguration
+            targetDataStore = WebKit::wrapper(dataStore);
+    });
+    if (targetDataStore)
+        [configuration setWebsiteDataStore:targetDataStore];
```
`targetDataStore.get()` once you switch to RetainPtr above.

the-chenergy committed:
Sites Logout Automatically After Brief Amount of Time
rdar://99829958
https://bugs.webkit.org/show_bug.cgi?id=245491

Reviewed by Sihui Liu.

Web Inspector currently uses its own default website data store, one
that comes with WKWebView with the default configurations. However,
this can cause conflicts when the owner browser already created a data
store pointing to the same file paths. As the browser may not expect
the default data store to be used at all, this patch makes Web Inspector
try to use the same custom data store object as the browser, avoiding
the conflicts.

I was not able reproduce the issue from the bug report, but I tested
that this patch still preserved Web Inspector's storage as normal.
I could still add breakpoints and change settings and have Web Inspector
remember those states across relaunches of Safari or the Minibrowser,
all while my code changes were executed.

* Source/WebKit/UIProcess/Inspector/mac/WKInspectorViewController.mm:
(-[WKInspectorViewController webViewConfiguration]):
   - When constructing the configurations to create a WKWebView, try
     searching for the data store object the browser would have created.

Canonical link: https://commits.webkit.org/286436@main
